README.md
# stactools-cop-dem

- Name: cop-dem
- Package: `stactools.cop_dem`
- PyPI: https://pypi.org/project/stactools-cop-dem/
- Owner: @justinfisk
- Dataset homepages:
  - https://copernicus-dem-30m.s3.amazonaws.com/readme.html
  - https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.032021.4326.1
- STAC extensions used:
  - [proj](https://github.com/stac-extensions/projection/)
  - [item-assets](https://github.com/stac-extensions/item-assets)
  - [raster](https://github.com/stac-extensions/raster)
- Extra fields: none

The Copernicus DEM is a Digital Surface Model (DSM) which represents the surface of the Earth including buildings, infrastructure and vegetation. We provide two instances of Copernicus DEM named GLO-30 Public and GLO-90. GLO-90 provides worldwide coverage at 90 meters. GLO-30 Public provides limited worldwide coverage at 30 meters because a small subset of tiles covering specific countries are not yet released to the public by the Copernicus Programme. Note that in both cases ocean areas do not have tiles, there one can assume height values equal to zero. Data is provided as Cloud Optimized GeoTIFFs.

(from https://copernicus-dem-30m.s3.amazonaws.com/readme.html )

## Examples

### STAC objects

- [Item](examples/Copernicus_DSM_COG_30_N53_00_W115_00_DEM.json)
- [Collection](examples/cop-dem-glo-30.json)

### Command-line usage

Description of the command line functions

```bash
stac cop-dem create-item Copernicus_DSM_COG_30_N53_00_W115_00_DEM.tif Copernicus_DSM_COG_30_N53_00_W115_00_DEM.json
```

```bash
stac cop-dem create-collection glo-30 . --host AWS
```

Use `stac cop-dem --help` to see all subcommands and options.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.

## [Unreleased]

- Add additional assets that are included in the data distribution: Masks and Previews. ([#6](https://github.com/stactools-packages/cop-dem/issues/6))

## [0.4.1] - 2023-04-14

### Changed

- Round `proj:centroid` values to one decimal place, since they'll always be in 0.5 degree increments([#28](https://github.com/stactools-packages/cop-dem/pull/28))

## [0.4.0] - 2023-04-14

## Added

- Added support for Grid Extension with property `grid:code` ([#24](https://github.com/stactools-packages/cop-dem/pull/24))
- Added Projection Extension property `proj:centroid` ([#26](https://github.com/stactools-packages/cop-dem/pull/26))

## [0.3.0] - 2023-03-27

## Changed

- The asset title is now always `Data`, rather than being specific to each item derived from the item ID. ([#21](https://github.com/stactools-packages/cop-dem/pull/21))

## [0.2.0] - 2023-03-24

## Changed

- Item `platform` value from `TanDEM-X` to `tandem-x` ([#15](https://github.com/stactools-packages/cop-dem/pull/15))

## Fixed

- Fixed license link to a valid URL ([#17](https://github.com/stactools-packages/cop-dem/pull/17))
- Item now propertly advertises use of the Raster Extension via stac_extensions ([#18](https://github.com/stactools-packages/cop-dem/pull/18))

## [0.1.1] - 2023-01-30

### Added

- Package description ([#10](https://github.com/stactools-packages/cop-dem/pull/10))

## [0.1.0] - 2023-01-30

### Added

- create-collection and matching tests for both glo-30 and glo-90
- New: `--host` option to specify a Provider HOST from a list of options in constants. `AWS` and `OT` are the current valid options.
- added python 3.10

### Removed

- remove python 3.7 support

[Unreleased]: https://github.com/stactools-packages/cop-dem/compare/v0.4.1..main
[0.4.1]: https://github.com/stactools-packages/cop-dem/compare/v0.4.0..v0.4.1
[0.4.0]: https://github.com/stactools-packages/cop-dem/compare/v0.3.0..v0.4.0
[0.3.0]: https://github.com/stactools-packages/cop-dem/compare/v0.2.0..v0.3.0
[0.2.0]: https://github.com/stactools-packages/cop-dem/compare/v0.1.1..v0.2.0
[0.1.1]: https://github.com/stactools-packages/cop-dem/compare/v0.1.0..v0.1.1
[0.1.0]: https://github.com/stactools-packages/cop-dem/tags/v0.1.0


tests/__init__.py
from stactools.testing import TestData

test_data = TestData(
    __file__,
    {
        "Copernicus_DSM_COG_10_N53_00_W115_00_DEM.tif": {
            "url": "s3://raster/COP30/COP30_hh/Copernicus_DSM_COG_10_N53_00_W115_00_DEM.tif",
            "s3": {
                "anon": True,
                "client_kwargs": {
                    "endpoint_url": "https://opentopography.s3.sdsc.edu",
                },
            },
        },
        "Copernicus_DSM_COG_30_N53_00_W115_00_DEM.tif": {
            "url": "s3://raster/COP90/COP90_hh/Copernicus_DSM_COG_30_N53_00_W115_00_DEM.tif",
            "s3": {
                "anon": True,
                "client_kwargs": {
                    "endpoint_url": "https://opentopography.s3.sdsc.edu",
                },
            },
        },
    },
)


tests/test_stac.py
import datetime
from unittest import TestCase

from pystac import Provider, MediaType
from pystac.extensions.grid import GridExtension
from pystac.extensions.projection import ProjectionExtension
from pystac.extensions.raster import RasterExtension
from pystac.provider import ProviderRole

from stactools.cop_dem import stac

from tests import test_data


class StacTest(TestCase):

    def setUp(self):
        self.glo30_path = test_data.get_external_data(
            "Copernicus_DSM_COG_10_N53_00_W115_00_DEM.tif")
        self.glo90_path = test_data.get_external_data(
            "Copernicus_DSM_COG_30_N53_00_W115_00_DEM.tif")

    def test_create_glo30_item(self):
        item = stac.create_item(self.glo30_path)
        self.assertEqual(item.id, "Copernicus_DSM_COG_10_N53_00_W115_00_DEM")
        self.assertIsNotNone(item.geometry)
        self.assertEqual(list(item.bbox), [
            -115.00020833333333, 53.00013888888889, -114.00020833333333,
            54.00013888888889
        ])
        self.assertEqual(
            item.datetime,
            datetime.datetime(2021, 4, 22, tzinfo=datetime.timezone.utc))

        common_metadata = item.common_metadata
        self.assertEqual(common_metadata.platform, "tandem-x")
        self.assertEqual(common_metadata.gsd, 30)
        expected_providers = [
            Provider("European Space Agency",
                     roles=[ProviderRole.LICENSOR],
                     url=("https://spacedata.copernicus.eu/documents/20126/0/"
                          "CSCDA_ESA_Mission-specific+Annex.pdf")),
            Provider("Sinergise",
                     roles=[ProviderRole.PRODUCER, ProviderRole.PROCESSOR],
                     url="https://registry.opendata.aws/copernicus-dem/"),
            Provider("OpenTopography",
                     roles=[ProviderRole.HOST],
                     url=("https://portal.opentopography.org/"
                          "datasetMetadata?otCollectionID=OT.032021.4326.1"))
        ]
        for expected, actual in zip(expected_providers,
                                    common_metadata.providers):
            self.assertDictEqual(expected.to_dict(), actual.to_dict())
        self.assertEqual(common_metadata.license, "proprietary")

        projection = ProjectionExtension.ext(item)
        self.assertEqual(projection.epsg, 4326)
        self.assertEqual(projection.shape, (3600, 2400))
        self.assertEqual(list(projection.transform), [
            0.00041666666666666664, 0.0, -115.00020833333333, 0.0,
            -0.0002777777777777778, 54.00013888888889
        ])

        grid = GridExtension.ext(item)
        self.assertEqual(grid.code, "CDEM-N53W115")

        handbook = item.get_single_link("handbook")
        self.assertIsNotNone(handbook)
        self.assertEqual(handbook.title, "Copernicus DEM User handbook")
        self.assertEqual(handbook.rel, "handbook")
        self.assertEqual(
            handbook.href,
            "https://object.cloud.sdsc.edu/v1/AUTH_opentopography/www/metadata"
            "/Copernicus_metadata.pdf")
        self.assertEqual(handbook.media_type, "application/pdf")

        data = item.assets["data"]
        self.assertEqual(data.href, self.glo30_path)
        self.assertEqual(data.title, "Data")
        self.assertIsNone(data.description)
        self.assertEqual(data.media_type, MediaType.COG)
        self.assertEqual(data.roles, ["data"])

        self.assertTrue(GridExtension.has_extension(item))
        self.assertTrue(ProjectionExtension.has_extension(item))
        self.assertTrue(RasterExtension.has_extension(item))

        item.validate()  # raises STACValidationError if not

    def test_create_glo90_item(self):
        item = stac.create_item(self.glo90_path)
        self.assertEqual(item.id, "Copernicus_DSM_COG_30_N53_00_W115_00_DEM")
        self.assertIsNotNone(item.geometry)
        self.assertEqual(
            list(item.bbox),
            [-115.000625, 53.000416666666666, -114.000625, 54.000416666666666])
        self.assertEqual(
            item.datetime,
            datetime.datetime(2021, 4, 22, tzinfo=datetime.timezone.utc))

        common_metadata = item.common_metadata
        self.assertEqual(common_metadata.platform, "tandem-x")
        self.assertEqual(common_metadata.gsd, 90)
        expected_providers = [
            Provider("European Space Agency",
                     roles=[ProviderRole.LICENSOR],
                     url=("https://spacedata.copernicus.eu/documents/20126/0/"
                          "CSCDA_ESA_Mission-specific+Annex.pdf")),
            Provider("Sinergise",
                     roles=[ProviderRole.PRODUCER, ProviderRole.PROCESSOR],
                     url="https://registry.opendata.aws/copernicus-dem/"),
            Provider("OpenTopography",
                     roles=[ProviderRole.HOST],
                     url=("https://portal.opentopography.org/"
                          "datasetMetadata?otCollectionID=OT.032021.4326.1"))
        ]
        for expected, actual in zip(expected_providers,
                                    common_metadata.providers):
            self.assertDictEqual(expected.to_dict(), actual.to_dict())
        self.assertEqual(common_metadata.license, "proprietary")

        projection = ProjectionExtension.ext(item)
        self.assertEqual(projection.epsg, 4326)
        self.assertEqual(projection.shape, (1200, 800))
        self.assertEqual(list(projection.transform), [
            0.00125, 0.0, -115.000625, 0.0, -0.0008333333333333334,
            54.000416666666666
        ])
        self.assertEqual(round(projection.centroid["lat"], 5), 53.5)
        self.assertEqual(round(projection.centroid["lon"], 5), -114.5)

        grid = GridExtension.ext(item)
        self.assertEqual(grid.code, "CDEM-N53W115")

        handbook = item.get_single_link("handbook")
        self.assertIsNotNone(handbook)
        self.assertEqual(handbook.title, "Copernicus DEM User handbook")
        self.assertEqual(handbook.rel, "handbook")
        self.assertEqual(
            handbook.href,
            "https://object.cloud.sdsc.edu/v1/AUTH_opentopography/www/metadata"
            "/Copernicus_metadata.pdf")
        self.assertEqual(handbook.media_type, "application/pdf")

        data = item.assets["data"]
        self.assertEqual(data.href, self.glo90_path)
        self.assertEqual(data.title, "Data")
        self.assertIsNone(data.description)
        self.assertEqual(data.media_type, MediaType.COG)
        self.assertEqual(data.roles, ["data"])

        self.assertTrue(ProjectionExtension.has_extension(item))
        self.assertTrue(RasterExtension.has_extension(item))
        self.assertTrue(GridExtension.has_extension(item))

        item.validate()  # raises STACValidationError if not

    def test_create_item_with_read_href_modifier(self):
        done = False

        def do_it(href):
            nonlocal done
            done = True
            return href

        _ = stac.create_item(self.glo30_path, read_href_modifier=do_it)
        self.assertTrue(done, "Didn't do it")

    def test_create_collection(self):
        # Write tests for each for the creation of a STAC Collection
        # Create the STAC Collection...
        collection = stac.create_collection("glo-30")
        collection.set_self_href("")

        # Check that it has some required attributes
        self.assertEqual(collection.id, "cop-dem-glo-30")
        # self.assertEqual(collection.other_attr...

        # Validate
        collection.validate()


tests/test_module.py
import unittest

import stactools.cop_dem


class TestModule(unittest.TestCase):
    def test_version(self):
        self.assertIsNotNone(stactools.cop_dem.__version__)


scripts/create_examples.py
#!/usr/bin/env python3

from pathlib import Path

from pystac import CatalogType

from stactools.cop_dem import stac

root = Path(__file__).parents[1]
examples = root / "examples"
tiles = root / "tests" / "data-files" / "external"


def gen_collection_example(collection_name):
    collection = stac.create_collection(collection_name)
    aa = int(str.split(collection_name, "-")[1]) // 3
    items = [stac.create_item([str(p) for p in tiles.glob(f"*{aa}*.tif")][0])]
    collection.add_items(items)

    collection.normalize_hrefs(str(examples))
    collection.make_all_asset_hrefs_relative()
    json_path = examples / f"cop-dem-{collection_name}"
    collection.save(
        dest_href=json_path.as_posix(), catalog_type=CatalogType.SELF_CONTAINED
    )


gen_collection_example("glo-30")
gen_collection_example("glo-90")


src/stactools/cop_dem/stac.py
import os.path
import re
from typing import Optional

from pystac import (
    CatalogType,
    Collection,
    Extent,
    Asset,
    Summaries,
    SpatialExtent,
    TemporalExtent,
)
from pystac.extensions.grid import GridExtension
from pystac.extensions.projection import ProjectionExtension
from pystac.extensions.item_assets import ItemAssetsExtension
from pystac.extensions.raster import (
    DataType,
    RasterBand,
    RasterExtension,
    Sampling,
)
from pystac.media_type import MediaType

import rasterio
from shapely.geometry import mapping, box, shape as make_shape
from pystac import Item

from stactools.core.io import ReadHrefModifier
from stactools.cop_dem import constants as co

from stactools.cop_dem.assets_utils import change_asset_directory


def create_item(
    href: str,
    read_href_modifier: Optional[ReadHrefModifier] = None,
    host: Optional[str] = None,
) -> Item:
    """Creates a STAC Item from a single tile of Copernicus DEM data."""
    if read_href_modifier:
        modified_href = read_href_modifier(href)
    else:
        modified_href = href
    with rasterio.open(modified_href) as dataset:
        if dataset.crs.to_epsg() != co.COP_DEM_EPSG:
            raise ValueError(f"Dataset {href} is not EPSG:{co.COP_DEM_EPSG}, "
                             "which is required for Copernicus DEM data")
        bbox = list(dataset.bounds)
        geometry = mapping(box(*bbox))
        transform = dataset.transform
        shape = dataset.shape
        item = Item(
            id=os.path.splitext(os.path.basename(href))[0],
            geometry=geometry,
            bbox=bbox,
            datetime=co.COP_DEM_COLLECTION_START,
            properties={},
            stac_extensions=[],
        )

    # resolution in arc seconds (not meters!), which is and 30 for GLO-90 and 10 for GLO-30
    p = re.compile(
        r"Copernicus_DSM_COG_(?P<res>\d{2})_(?P<northing>[NS]\d{2})_00_(?P<easting>[EW]\d{3})_00_DEM.*"  # noqa: E501
    )
    if m := p.match(os.path.basename(href)):
        res = m.group("res")
        if res == "30":
            gsd = 90
        elif res == "10":
            gsd = 30
        else:
            raise ValueError(f"unknown resolution {res}")

        northing = m.group("northing")
        easting = m.group("easting")
    else:
        raise ValueError(f"unable to parse {href}")

    item.add_links(co.COP_DEM_LINKS)
    item.common_metadata.platform = co.COP_DEM_PLATFORM
    item.common_metadata.gsd = gsd

    if host and host not in co.COP_DEM_HOST.keys():
        raise ValueError(
            f"Invalid host: {host}. Must be one of {list(co.COP_DEM_HOST.keys())}"
        )
    if host and (host_provider := co.COP_DEM_HOST.get(host)):
        providers = [*co.COP_DEM_PROVIDERS, host_provider]
    else:
        providers = co.COP_DEM_PROVIDERS

    item.common_metadata.providers = providers
    item.common_metadata.license = "proprietary"

    data_asset = Asset(
        href=href,
        title="Data",
        description=None,
        media_type=MediaType.COG,
        roles=["data"],
    )
    data_bands = RasterBand.create(
        sampling=Sampling.POINT,
        data_type=DataType.FLOAT32,
        spatial_resolution=gsd,
        unit="meter",
    )

    item.add_asset("data", data_asset)
    RasterExtension.ext(data_asset, add_if_missing=True).bands = [data_bands]

    # Make a list of /AUXFILES assets
    assets_auxfiles = ["EDM.tif", "FLM.tif", "WBM.tif", "HEM.tif", "ACM.kml"]
    assets_auxfiles_dict = dict([
        change_asset_directory(href=href,
                               asset_name=asset,
                               subdirectory="AUXFILES")
        for asset in assets_auxfiles
    ])

    # Make a list of /PREVIEW assets
    assets_previews = [
        "SRC.kml", "DEM_QL.tif", "QL.kml", "DEM_ABS_QL.tif", "EDM_QL.tif",
        "FLM_QL.tif", "WBM_QL.tif", "HEM_QL.tif"
    ]
    assets_dict = dict([
        change_asset_directory(href=href,
                               asset_name=asset,
                               subdirectory="PREVIEW")
        for asset in assets_previews
    ])

    # Add the AUX and Preview Assets together
    assets_dict.update(assets_auxfiles_dict)

    for key, value in assets_dict.items():
        if (asset_def := co.COP_DEM_ASSETS.get(key)) is not None:
            item.add_asset(key, asset_def.create_asset(value))

    # Include the Classification Schema for Mask assets
    item.stac_extensions.append(co.CLASSIFICATION_EXTENSION_SCHEMA)

    projection = ProjectionExtension.ext(item, add_if_missing=True)
    projection.epsg = co.COP_DEM_EPSG
    projection.transform = transform[0:6]
    projection.shape = shape

    centroid = make_shape(item.geometry).centroid
    projection.centroid = {
        "lat": round(centroid.y, 1),
        "lon": round(centroid.x, 1)
    }

    grid = GridExtension.ext(item, add_if_missing=True)
    grid.code = f"CDEM-{northing}{easting}"

    return item


def create_collection(product: str, host: Optional[str] = None) -> Collection:
    """Create a STAC Collection

    This function includes logic to extract all relevant metadata from
    an asset describing the STAC collection and/or metadata coded into an
    accompanying constants.py file.

    See `Collection<https://pystac.readthedocs.io/en/latest/api.html#collection>`_.

    Args:
        Product (str): glo-30 or glo-90

    Returns:
        Item: STAC Item object

    Returns:
        Collection: STAC Collection object
    """
    if product.lower() == "glo-30":
        summaries = {
            "gsd": [30],
            "platform": [co.COP_DEM_PLATFORM],
            # "instruments": ,
        }
    elif product.lower() == "glo-90":
        summaries = {
            "gsd": [90],
            "platform": [co.COP_DEM_PLATFORM],
            # "instruments": ,
        }
    else:
        raise ValueError(
            f"{product} is not a valid product. Must be one of {co.COP_DEM_PRODUCTS}"
        )

    # Allow host to be selected by cli option
    if host and host not in co.COP_DEM_HOST.keys():
        raise ValueError(
            f"Invalid host: {host}. Must be one of {list(co.COP_DEM_HOST.keys())}"
        )
    if host and (host_provider := co.COP_DEM_HOST.get(host)):
        providers = [*co.COP_DEM_PROVIDERS, host_provider]
    else:
        providers = co.COP_DEM_PROVIDERS

    collection = Collection(
        id=f"cop-dem-{product.lower()}",
        title=f"Copernicus DEM {product.upper()}",
        description=co.COP_DEM_DESCRIPTION,
        license="proprietary",
        keywords=co.COP_DEM_KEYWORDS,
        catalog_type=CatalogType.RELATIVE_PUBLISHED,
        summaries=Summaries(summaries),
        extent=Extent(
            SpatialExtent(co.COP_DEM_SPATIAL_EXTENT),
            TemporalExtent([co.COP_DEM_TEMPORAL_EXTENT]),
        ),
        providers=providers,
        stac_extensions=[
            ItemAssetsExtension.get_schema_uri(),
            ProjectionExtension.get_schema_uri(),
            RasterExtension.get_schema_uri(),
            co.CLASSIFICATION_EXTENSION_SCHEMA
        ],
    )

    collection.add_links(co.COP_DEM_LINKS)

    assets = ItemAssetsExtension.ext(collection, add_if_missing=True)
    assets.item_assets = co.COP_DEM_ASSETS

    return collection


src/stactools/cop_dem/classes.py
# Classification list for Mask asset bands.

EDM_classes = [
    {
        'name': 'Void_no_data',
        'value': 0,
        'description': 'Void (no data)'
    },
    {
        'name': 'Not_edited',
        'value': 1,
        'description': 'Not edited'
    },
    {
        'name': 'Infill_of_external_elevation_data',
        'value': 2,
        'description': 'Infill of external elevation data'
    },
    {
        'name': 'Interpolated_pixels',
        'value': 3,
        'description': 'Interpolated pixels'
    },
    {
        'name': 'Smoothed_pixels',
        'value': 4,
        'description': 'Smoothed pixels'
    },
    {
        'name': 'Airport_editing',
        'value': 5,
        'description': 'Airport editing'
    },
    {
        'name': 'Raised_negative_elevation_pixels',
        'value': 6,
        'description': 'Raised negative elevation pixels'
    },
    {
        'name': 'Flattened_pixels',
        'value': 7,
        'description': 'Flattened pixels'
    },
    {
        'name': 'Ocean_pixels',
        'value': 8,
        'description': 'Ocean pixels'
    },
    {
        'name': 'Lake_pixels',
        'value': 9,
        'description': 'Lake pixels'
    },
    {
        'name': 'River_pixels',
        'value': 10,
        'description': 'River pixels'
    },
    {
        'name': 'Shoreline_pixels',
        'value': 11,
        'description': 'Shoreline pixels'
    },
    {
        'name': 'Morphed_pixels_(series_of_pixels_manually_set)',
        'value': 12,
        'description': 'Morphed pixels (series of pixels manually set)'
    },
    {
        'name': 'Shifted_pixels',
        'value': 13,
        'description': 'Shifted pixels'
    },
]

FLM_classes = [
    {
        'name': 'Void_no_data',
        'value': 0,
        'description': 'Void (no data)'
    },
    {
        'name': 'Edited_except_filled_pixels',
        'value': 1,
        'description': 'Edited (except filled pixels)'
    },
    {
        'name': 'Not_edited_not_filled',
        'value': 2,
        'description': 'Not edited / not filled'
    },
    {
        'name': 'ASTER',
        'value': 3,
        'description': 'ASTER'
    },
    {
        'name': 'SRTM90',
        'value': 4,
        'description': 'SRTM90'
    },
    {
        'name': 'SRTM30',
        'value': 5,
        'description': 'SRTM30'
    },
    {
        'name': 'GMTED2010',
        'value': 6,
        'description': 'GMTED2010'
    },
    {
        'name': 'SRTM30plus',
        'value': 7,
        'description': 'SRTM30plus'
    },
    {
        'name': 'TerraSAR-X_Radargrammetric_DEM',
        'value': 8,
        'description': 'TerraSAR-X Radargrammetric DEM'
    },
    {
        'name': 'AW3D306',
        'value': 9,
        'description': 'AW3D306'
    },
    {
        'name': 'Norway_DEM',
        'value': 100,
        'description': 'Norway DEM'
    },
    {
        'name': 'DSM05_Spain',
        'value': 101,
        'description': 'DSM05 Spain'
    },
    {
        'name': 'Norway_DEM_v2',
        'value': 102,
        'description': 'Norway DEM v2'
    },
]

WBM_classes = [
    {
        'name': 'No_water',
        'value': 0,
        'description': 'No water'
    },
    {
        'name': 'Ocean',
        'value': 1,
        'description': 'Ocean'
    },
    {
        'name': 'Lake',
        'value': 2,
        'description': 'Lake'
    },
    {
        'name': 'River',
        'value': 3,
        'description': 'River'
    },
]


src/stactools/cop_dem/commands.py
import os

import click

from stactools.cop_dem import constants as co
from stactools.cop_dem import stac


def create_cop_dem_command(cli):
    """Creates the COP DEM stactools commands."""

    @cli.group("cop-dem", short_help="Work with Copernicus DEM data.")
    def cop_dem():
        pass

    @cop_dem.command(
        "create-item", short_help="Creates a STAC item from an Copernicus DEM tile."
    )
    @click.argument("source")
    @click.argument("destination")
    @click.option(
        "--validate/--no-validate", default=True, help="Validate the item before saving"
    )
    @click.option(
        "--host",
        type=click.Choice(co.COP_DEM_HOST.keys(), case_sensitive=False),
        default=None,
        help="Set PROVIDER HOST",
    )
    def create_item_command(source: str, destination: str, validate: bool, host: str):
        """Creates a STAC Item

        source: Path to input item

        destination: File to output json

        """
        item = stac.create_item(source, host=host)
        if validate:
            item.validate()
        item.save_object(dest_href=destination)

    @cop_dem.command(
        "create-collection", short_help="Creates a STAC Collection for Copernicus DEM."
    )
    @click.argument("product", type=click.Choice(["glo-30", "glo-90"]))
    @click.argument("destination", type=click.Path(exists=True))
    @click.option(
        "-u", "--url", default="", type=str, help="Root url to prepend to all records"
    )
    @click.option(
        "--validate/--no-validate", default=True, help="Validate the item before saving"
    )
    @click.option(
        "--host",
        type=click.Choice(co.COP_DEM_HOST.keys(), case_sensitive=False),
        default=None,
        help="Set PROVIDER HOST",
    )
    def create_collection_command(
        destination: str, product: str, url: str, validate: bool, host: str
    ):
        """Creates a STAC Collection

        product: The DEM product, glo-30 or glo-90

        destination: Folder to output json

        """
        collection = stac.create_collection(product, host=host)
        json_path = os.path.join(destination, f"{collection.id}.json")
        collection.set_self_href(
            os.path.join(url, collection.id, os.path.basename(json_path))
        )
        if validate:
            collection.validate()
        collection.save_object(dest_href=json_path)

    return cop_dem


src/stactools/cop_dem/constants.py
from datetime import datetime
from typing import Optional

from pystac import Provider, Link, ProviderRole
from pystac.extensions.item_assets import AssetDefinition
from pystac.extensions.raster import DataType
from pystac.utils import str_to_datetime
from pystac.media_type import MediaType
from stactools.cop_dem.classes import EDM_classes, FLM_classes, WBM_classes

CLASSIFICATION_EXTENSION_SCHEMA = "https://stac-extensions.github.io/classification/v1.1.0/schema.json"  # noqa:E501

UINT8_BAND = [{"data_type": DataType.UINT8}]

FLOAT32_BAND = [{"data_type": DataType.FLOAT32}]

COP_DEM_PRODUCTS = ['glo-30', 'glo-90']

COP_DEM_SPATIAL_EXTENT = [[-180., -90., 180., 90.]]

COP_DEM_COLLECTION_START: Optional[datetime] = str_to_datetime(
    "2021-04-22T00:00:00Z")

COP_DEM_COLLECTION_END: Optional[datetime] = str_to_datetime(
    "2021-04-22T00:00:00Z")

COP_DEM_TEMPORAL_EXTENT = [COP_DEM_COLLECTION_START,
                           COP_DEM_COLLECTION_END]  # TODO: find the dates

COP_DEM_PLATFORM = "tandem-x"

COP_DEM_EPSG = 4326

COP_DEM_KEYWORDS = ['DEM', 'COPERNICUS', 'DSM', 'Elevation']

COP_DEM_PROVIDERS = [
    Provider("European Space Agency",
             roles=[ProviderRole.LICENSOR],
             url=("https://spacedata.copernicus.eu/documents/20126/0/"
                  "CSCDA_ESA_Mission-specific+Annex.pdf")),
    Provider("Sinergise",
             roles=[ProviderRole.PRODUCER, ProviderRole.PROCESSOR],
             url="https://registry.opendata.aws/copernicus-dem/"),
]

# Various HOST options
COP_DEM_HOST = {
    "OT":
    Provider("OpenTopography",
             roles=[ProviderRole.HOST],
             url=("https://portal.opentopography.org/"
                  "datasetMetadata?otCollectionID=OT.032021.4326.1")),
    "AWS":
    Provider("Sinergise",
             roles=[ProviderRole.HOST],
             url="https://registry.opendata.aws/copernicus-dem/")
}

COP_DEM_LINKS = [
    Link(
        "handbook",
        "https://object.cloud.sdsc.edu/v1/AUTH_opentopography/www/metadata/Copernicus_metadata.pdf",
        "application/pdf",
        "Copernicus DEM User handbook",
        extra_fields={"description": "Also includes data usage information"}),
    Link(
        "product handbook",
        "https://spacedata.copernicus.eu/documents/20123/122407/GEO1988-CopernicusDEM-SPE-002_ProductHandbook_I5.0+%281%29.pdf/706ee17d-2cce-f1fa-a73e-1686d28f09dd?t=1679657087883"  # noqa: E501
        "application/pdf",
        "Copernicus DEM Product Handbook (Nov 2022)",
        extra_fields={"description": "Also includes data usage information"}),
    Link(
        "license",
        "https://spacedata.copernicus.eu/documents/20123/121286/CSCDA_ESA_Mission-specific+Annex_31_Oct_22.pdf",  # noqa: E501
        "Copernicus Data Access")
]

COP_DEM_ASSETS = {
    "data":
    AssetDefinition({
        "type": MediaType.COG,
        "roles": ["data"],
    }),
    "edm":
    AssetDefinition({
        "title": "Editing Mask",
        "type": MediaType.COG,
        "description": "Editing Mask",
        "role": ["mask"],
        "raster:bands": UINT8_BAND,
        "classification:classes": EDM_classes,
    }),
    "flm":
    AssetDefinition({
        "title": "Filling Mask",
        "type": MediaType.COG,
        "description": "Filling Mask",
        "role": ["mask"],
        "raster:bands": UINT8_BAND,
        "classification:classes": FLM_classes,
    }),
    "wbm":
    AssetDefinition({
        "title": "Water Body Mask",
        "type": MediaType.COG,
        "description": "Water Body Mask",
        "role": ["mask"],
        "raster:bands": UINT8_BAND,
        "classification:classes": WBM_classes,
    }),
    "hem":
    AssetDefinition({
        "title": "Height Error Mask",
        "type": MediaType.COG,
        "description": "Height Error Mask",
        "role": ["mask"],
        "raster:bands": FLOAT32_BAND,
    }),
    "acm":
    AssetDefinition({
        "title": "Accuracy Layer",
        "type": "application/vnd.google-earth.kml+xml",
        "description": "Accuracy Layer",
        "role": ["mask"],
    }),
    "src":
    AssetDefinition({
        "title": "Source Mask",
        "type": "application/vnd.google-earth.kml+xml",
        "description": "Source Mask",
        "role": ["mask"],
    }),
    "dem_ql":
    AssetDefinition({
        "title": "DEM QuickLook Relative",
        "type": MediaType.GEOTIFF,
        "description":
        "QuickLook of Digital Elevation Model with shaded relief representation (relative)",
        "role": ["overview"],
    }),
    "ql":
    AssetDefinition({
        "title": "QuickLook KML",
        "type": "application/vnd.google-earth.kml+xml",
        "description": "KML of QuickLook data visualization",
        "role": ["overview"],
    }),
    "dem_abs_ql":
    AssetDefinition({
        "title": "DEM QuickLook Absolute",
        "type": MediaType.GEOTIFF,
        "description":
        "QuickLook of Digital Elevation Model with shaded relief representation (absolute)",
        "role": ["overview"],
    }),
    "edm_ql":
    AssetDefinition({
        "title": "Editing Mask QuickLook",
        "type": MediaType.GEOTIFF,
        "description": "QuickLook of Editing Mask",
        "role": ["overview"],
    }),
    "flm_ql":
    AssetDefinition({
        "title": "Filling Mask QuickLook",
        "type": MediaType.GEOTIFF,
        "description": "QuickLook of Filling Mask",
        "role": ["overview"],
    }),
    "wbm_ql":
    AssetDefinition({
        "title": "Water Body Mask QuickLook",
        "type": MediaType.GEOTIFF,
        "description": "QuickLook of Water Body Mask",
        "role": ["overview"],
    }),
    "hem_ql":
    AssetDefinition({
        "title": "Height Error Mask QuickLook",
        "type": MediaType.GEOTIFF,
        "description": "QuickLook of Height Error Mask",
        "role": ["overview"],
    }),
}

COP_DEM_DESCRIPTION = '''The Copernicus DEM is a Digital Surface Model (DSM) which represents the surface of the Earth including buildings, infrastructure and vegetation. We provide two instances of Copernicus DEM named GLO-30 Public and GLO-90. GLO-90 provides worldwide coverage at 90 meters. GLO-30 Public provides limited worldwide coverage at 30 meters because a small subset of tiles covering specific countries are not yet released to the public by the Copernicus Programme. Note that in both cases ocean areas do not have tiles, there one can assume height values equal to zero. Data is provided as Cloud Optimized GeoTIFFs and comes from Copernicus DEM 2021 release.'''  # noqa: E501
